File: src/main/java/org/example/kalkulationsprogramm/service/FormularTextbausteinDefaultService.java
package org.example.kalkulationsprogramm.service;

import java.util.ArrayList;
import java.util.EnumMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

import org.example.kalkulationsprogramm.domain.Dokumenttyp;
import org.example.kalkulationsprogramm.domain.FormularTemplateTextbausteinDefault;
import org.example.kalkulationsprogramm.domain.Textbaustein;
import org.example.kalkulationsprogramm.domain.TextbausteinPosition;
import org.example.kalkulationsprogramm.dto.Formular.FormularTextbausteinDefaultsDto;
import org.example.kalkulationsprogramm.repository.FormularTemplateTextbausteinDefaultRepository;
import org.example.kalkulationsprogramm.repository.TextbausteinRepository;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.StringUtils;
import org.springframework.web.server.ResponseStatusException;

import lombok.RequiredArgsConstructor;

/**
 * Verwaltet Default-Textbausteine, die beim Anlegen oder Umwandeln eines
 * Geschaeftsdokuments automatisch vor bzw. nach den Leistungen eingefuegt werden.
 * Konfiguriert wird das pro Vorlage (Briefpapier) und Dokumenttyp im FormularwesenEditor.
 */
@Service
@RequiredArgsConstructor
public class FormularTextbausteinDefaultService {

    private final FormularTemplateTextbausteinDefaultRepository defaultsRepository;
    private final TextbausteinRepository textbausteinRepository;

    /** Liefert alle Defaults einer Vorlage als DTO, gruppiert pro Dokumenttyp. */
    @Transactional(readOnly = true)
    public FormularTextbausteinDefaultsDto loadDefaults(String templateName) {
        FormularTextbausteinDefaultsDto dto = new FormularTextbausteinDefaultsDto();
        if (!StringUtils.hasText(templateName)) {
            return dto;
        }
        List<FormularTemplateTextbausteinDefault> rows =
                defaultsRepository.findByTemplateNameIgnoreCaseOrderByDokumenttypAscPositionAscSortOrderAsc(
                        templateName.trim());

        // Gruppiert: Dokumenttyp -> (Position -> Liste<Long>)
        Map<Dokumenttyp, EnumMap<TextbausteinPosition, List<Long>>> grouped = new EnumMap<>(Dokumenttyp.class);
        for (FormularTemplateTextbausteinDefault row : rows) {
            if (row.getTextbaustein() == null || row.getTextbaustein().getId() == null) continue;
            grouped
                    .computeIfAbsent(row.getDokumenttyp(), k -> new EnumMap<>(TextbausteinPosition.class))
                    .computeIfAbsent(row.getPosition(), k -> new ArrayList<>())
                    .add(row.getTextbaustein().getId());
        }

        grouped.forEach((typ, byPos) -> {
            FormularTextbausteinDefaultsDto.Entry entry = new FormularTextbausteinDefaultsDto.Entry();
            entry.setDokumenttyp(typ.getLabel());
            entry.setVortextIds(byPos.getOrDefault(TextbausteinPosition.VOR, List.of()));
            entry.setNachtextIds(byPos.getOrDefault(TextbausteinPosition.NACH, List.of()));
            dto.getEntries().add(entry);
        });

        return dto;
    }

    /** Ersetzt alle Defaults einer Vorlage durch die im DTO angegebenen Werte. */
    @Transactional
    public void replaceDefaults(String templateName, FormularTextbausteinDefaultsDto dto) {
        if (!StringUtils.hasText(templateName)) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Vorlagenname fehlt.");
        }
        String name = templateName.trim();
        defaultsRepository.deleteByTemplateNameIgnoreCase(name);
        if (dto == null || dto.getEntries() == null) {
            return;
        }

        for (FormularTextbausteinDefaultsDto.Entry entry : dto.getEntries()) {
            if (entry == null || !StringUtils.hasText(entry.getDokumenttyp())) continue;
            Dokumenttyp typ = Dokumenttyp.fromLabel(entry.getDokumenttyp());
            saveAll(name, typ, TextbausteinPosition.VOR, entry.getVortextIds());
            saveAll(name, typ, TextbausteinPosition.NACH, entry.getNachtextIds());
        }
    }

    /** Wird beim Loeschen oder Umbenennen einer Vorlage aufgerufen. */
    @Transactional
    public void deleteForTemplate(String templateName) {
        if (!StringUtils.hasText(templateName)) return;
        defaultsRepository.deleteByTemplateNameIgnoreCase(templateName.trim());
    }

    /**
     * Kopiert alle Defaults einer Vorlage auf einen neuen Namen
     * (z.B. nach Rename oder Copy einer Briefpapier-Vorlage).
     */
    @Transactional
    public void renameTemplate(String oldName, String newName) {
        if (!StringUtils.hasText(oldName) || !StringUtils.hasText(newName)) return;
        if (oldName.equalsIgnoreCase(newName)) return;

        List<FormularTemplateTextbausteinDefault> rows =
                defaultsRepository.findByTemplateNameIgnoreCaseOrderByDokumenttypAscPositionAscSortOrderAsc(oldName.trim());
        defaultsRepository.deleteByTemplateNameIgnoreCase(newName.trim());
        for (FormularTemplateTextbausteinDefault row : rows) {
            FormularTemplateTextbausteinDefault copy = new FormularTemplateTextbausteinDefault();
            copy.setTemplateName(newName.trim());
            copy.setDokumenttyp(row.getDokumenttyp());
            copy.setPosition(row.getPosition());
            copy.setTextbaustein(row.getTextbaustein());
            copy.setSortOrder(row.getSortOrder());
            defaultsRepository.save(copy);
        }
        defaultsRepository.deleteByTemplateNameIgnoreCase(oldName.trim());
    }

    /**
     * Liefert die Default-Textbausteine fuer einen konkreten Dokumenttyp einer Vorlage,
     * mit den vollstaendigen Textbaustein-Objekten in der definierten Reihenfolge.
     * Wird vom DocumentEditor beim Neuanlegen / Umwandeln aufgerufen.
     */
    @Transactional(readOnly = true)
    public DefaultsForDokumenttyp loadForDokumenttyp(String templateName, String dokumenttyp) {
        if (!StringUtils.hasText(templateName) || !StringUtils.hasText(dokumenttyp)) {
            return new DefaultsForDokumenttyp(List.of(), List.of());
        }
        Dokumenttyp typ;
        try {
            typ = Dokumenttyp.fromLabel(dokumenttyp.trim());
        } catch (IllegalArgumentException e) {
            return new DefaultsForDokumenttyp(List.of(), List.of());
        }

        List<Textbaustein> vor = defaultsRepository
                .findByTemplateNameIgnoreCaseAndDokumenttypAndPositionOrderBySortOrderAsc(
                        templateName.trim(), typ, TextbausteinPosition.VOR)
                .stream().map(FormularTemplateTextbausteinDefault::getTextbaustein)
                .filter(Objects::nonNull).toList();

        List<Textbaustein> nach = defaultsRepository
                .findByTemplateNameIgnoreCaseAndDokumenttypAndPositionOrderBySortOrderAsc(
                        templateName.trim(), typ, TextbausteinPosition.NACH)
                .stream().map(FormularTemplateTextbausteinDefault::getTextbaustein)
                .filter(Objects::nonNull).toList();

        return new DefaultsForDokumenttyp(vor, nach);
    }

    private void saveAll(String templateName, Dokumenttyp typ, TextbausteinPosition pos, List<Long> ids) {
        if (ids == null || ids.isEmpty()) return;
        int order = 0;
        for (Long id : ids) {
            if (id == null) continue;
            Textbaustein tb = textbausteinRepository.findById(id).orElse(null);
            if (tb == null) continue; // unbekannte ID stillschweigend ueberspringen
            FormularTemplateTextbausteinDefault row = new FormularTemplateTextbausteinDefault();
            row.setTemplateName(templateName);
            row.setDokumenttyp(typ);
            row.setPosition(pos);
            row.setTextbaustein(tb);
            row.setSortOrder(order++);
            defaultsRepository.save(row);
        }
    }

    public record DefaultsForDokumenttyp(List<Textbaustein> vortexte, List<Textbaustein> nachtexte) {}
}

File: src/main/kotlin/org/example/kalkulationsprogramm/service/FormularTextbausteinDefaultService.kt
package org.example.kalkulationsprogramm.service

import org.example.kalkulationsprogramm.domain.Dokumenttyp
import org.example.kalkulationsprogramm.domain.FormularTemplateTextbausteinDefault
import org.example.kalkulationsprogramm.domain.Textbaustein
import org.example.kalkulationsprogramm.domain.TextbausteinPosition
import org.example.kalkulationsprogramm.dto.Formular.FormularTextbausteinDefaultsDto
import org.example.kalkulationsprogramm.repository.FormularTemplateTextbausteinDefaultRepository
import org.example.kalkulationsprogramm.repository.TextbausteinRepository
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.util.StringUtils
import org.springframework.web.server.ResponseStatusException
import java.util.EnumMap

@Service
open class FormularTextbausteinDefaultService(
    private val defaultsRepository: FormularTemplateTextbausteinDefaultRepository,
    private val textbausteinRepository: TextbausteinRepository,
) {

    @Transactional(readOnly = true)
    open fun loadDefaults(templateName: String?): FormularTextbausteinDefaultsDto {
        val dto = FormularTextbausteinDefaultsDto()
        if (!StringUtils.hasText(templateName)) {
            return dto
        }
        val rows = defaultsRepository
            .findByTemplateNameIgnoreCaseOrderByDokumenttypAscPositionAscSortOrderAsc(templateName!!.trim())

        val grouped = EnumMap<Dokumenttyp, EnumMap<TextbausteinPosition, MutableList<Long>>>(Dokumenttyp::class.java)
        for (row in rows) {
            val textbausteinId = row.textbaustein?.id ?: continue
            val typ = row.dokumenttyp ?: continue
            val position = row.position ?: continue
            grouped
                .computeIfAbsent(typ) { EnumMap(TextbausteinPosition::class.java) }
                .computeIfAbsent(position) { mutableListOf() }
                .add(textbausteinId)
        }

        grouped.forEach { (typ, byPos) ->
            val entry = FormularTextbausteinDefaultsDto.Entry()
            entry.dokumenttyp = typ.label
            entry.vortextIds = byPos[TextbausteinPosition.VOR] ?: mutableListOf()
            entry.nachtextIds = byPos[TextbausteinPosition.NACH] ?: mutableListOf()
            dto.entries.add(entry)
        }

        return dto
    }

    @Transactional
    open fun replaceDefaults(templateName: String?, dto: FormularTextbausteinDefaultsDto?) {
        if (!StringUtils.hasText(templateName)) {
            throw ResponseStatusException(HttpStatus.BAD_REQUEST, "Vorlagenname fehlt.")
        }
        val name = templateName!!.trim()
        defaultsRepository.deleteByTemplateNameIgnoreCase(name)
        val entries = dto?.entries ?: return

        for (entry in entries) {
            if (!StringUtils.hasText(entry.dokumenttyp)) continue
            val typ = Dokumenttyp.fromLabel(entry.dokumenttyp) ?: continue
            saveAll(name, typ, TextbausteinPosition.VOR, entry.vortextIds)
            saveAll(name, typ, TextbausteinPosition.NACH, entry.nachtextIds)
        }
    }

    @Transactional
    open fun deleteForTemplate(templateName: String?) {
        if (!StringUtils.hasText(templateName)) return
        defaultsRepository.deleteByTemplateNameIgnoreCase(templateName!!.trim())
    }

    @Transactional
    open fun renameTemplate(oldName: String?, newName: String?) {
        if (!StringUtils.hasText(oldName) || !StringUtils.hasText(newName)) return
        if (oldName!!.equals(newName, ignoreCase = true)) return

        val oldTrimmed = oldName.trim()
        val newTrimmed = newName!!.trim()
        val rows = defaultsRepository
            .findByTemplateNameIgnoreCaseOrderByDokumenttypAscPositionAscSortOrderAsc(oldTrimmed)
        defaultsRepository.deleteByTemplateNameIgnoreCase(newTrimmed)
        for (row in rows) {
            val copy = FormularTemplateTextbausteinDefault()
            copy.templateName = newTrimmed
            copy.dokumenttyp = row.dokumenttyp
            copy.position = row.position
            copy.textbaustein = row.textbaustein
            copy.sortOrder = row.sortOrder
            defaultsRepository.save(copy)
        }
        defaultsRepository.deleteByTemplateNameIgnoreCase(oldTrimmed)
    }

    @Transactional(readOnly = true)
    open fun loadForDokumenttyp(templateName: String?, dokumenttyp: String?): DefaultsForDokumenttyp {
        if (!StringUtils.hasText(templateName) || !StringUtils.hasText(dokumenttyp)) {
            return DefaultsForDokumenttyp(emptyList(), emptyList())
        }
        val typ = try {
            Dokumenttyp.fromLabel(dokumenttyp!!.trim())
        } catch (_: IllegalArgumentException) {
            null
        } ?: return DefaultsForDokumenttyp(emptyList(), emptyList())

        val vor = defaultsRepository
            .findByTemplateNameIgnoreCaseAndDokumenttypAndPositionOrderBySortOrderAsc(
                templateName!!.trim(),
                typ,
                TextbausteinPosition.VOR,
            )
            .mapNotNull { it.textbaustein }

        val nach = defaultsRepository
            .findByTemplateNameIgnoreCaseAndDokumenttypAndPositionOrderBySortOrderAsc(
                templateName.trim(),
                typ,
                TextbausteinPosition.NACH,
            )
            .mapNotNull { it.textbaustein }

        return DefaultsForDokumenttyp(vor, nach)
    }

    private fun saveAll(templateName: String, typ: Dokumenttyp, pos: TextbausteinPosition, ids: List<Long>?) {
        if (ids.isNullOrEmpty()) return
        var order = 0
        for (id in ids) {
            val textbaustein = textbausteinRepository.findById(id).orElse(null) ?: continue
            val row = FormularTemplateTextbausteinDefault()
            row.templateName = templateName
            row.dokumenttyp = typ
            row.position = pos
            row.textbaustein = textbaustein
            row.sortOrder = order++
            defaultsRepository.save(row)
        }
    }

    data class DefaultsForDokumenttyp(
        val vortexte: List<Textbaustein>,
        val nachtexte: List<Textbaustein>,
    ) {
        fun vortexte(): List<Textbaustein> = vortexte

        fun nachtexte(): List<Textbaustein> = nachtexte
    }
}
